apps/list.c

```diff
@@ -606,29 +906,40 @@ static void list_pkey(void)
 {
     int i;
 
-    for (i = 0; i < EVP_PKEY_asn1_get_count(); i++) {
-        const EVP_PKEY_ASN1_METHOD *ameth;
-        int pkey_id, pkey_base_id, pkey_flags;
-        const char *pinfo, *pem_str;
-        ameth = EVP_PKEY_asn1_get0(i);
-        EVP_PKEY_asn1_get0_info(&pkey_id, &pkey_base_id, &pkey_flags,
-                                &pinfo, &pem_str, ameth);
-        if (pkey_flags & ASN1_PKEY_ALIAS) {
-            BIO_printf(bio_out, "Name: %s\n", OBJ_nid2ln(pkey_id));
-            BIO_printf(bio_out, "\tAlias for: %s\n",
-                       OBJ_nid2ln(pkey_base_id));
-        } else {
-            BIO_printf(bio_out, "Name: %s\n", pinfo);
-            BIO_printf(bio_out, "\tType: %s Algorithm\n",
-                       pkey_flags & ASN1_PKEY_DYNAMIC ?
-                       "External" : "Builtin");
-            BIO_printf(bio_out, "\tOID: %s\n", OBJ_nid2ln(pkey_id));
-            if (pem_str == NULL)
-                pem_str = "(none)";
-            BIO_printf(bio_out, "\tPEM string: %s\n", pem_str);
+    if (select_name == NULL) {
+        BIO_printf(bio_out, "Legacy:\n");
+        for (i = 0; i < EVP_PKEY_asn1_get_count(); i++) {
+            const EVP_PKEY_ASN1_METHOD *ameth;
+            int pkey_id, pkey_base_id, pkey_flags;
+            const char *pinfo, *pem_str;
+            ameth = EVP_PKEY_asn1_get0(i);
+            EVP_PKEY_asn1_get0_info(&pkey_id, &pkey_base_id, &pkey_flags,
+                                    &pinfo, &pem_str, ameth);
+            if (pkey_flags & ASN1_PKEY_ALIAS) {
+                BIO_printf(bio_out, " Name: %s\n", OBJ_nid2ln(pkey_id));
+                BIO_printf(bio_out, "\tAlias for: %s\n",
+                           OBJ_nid2ln(pkey_base_id));
+            } else {
+                BIO_printf(bio_out, " Name: %s\n", pinfo);
+                BIO_printf(bio_out, "\tType: %s Algorithm\n",
+                           pkey_flags & ASN1_PKEY_DYNAMIC ?
+                           "External" : "Builtin");
+                BIO_printf(bio_out, "\tOID: %s\n", OBJ_nid2ln(pkey_id));
+                if (pem_str == NULL)
+                    pem_str = "(none)";
+                BIO_printf(bio_out, "\tPEM string: %s\n", pem_str);
+            }
         }
-
     }
+    BIO_printf(bio_out, "Provided:\n");
+    BIO_printf(bio_out, " Key encryption:\n");
+    list_asymciphers();
+    BIO_printf(bio_out, " Key Exchange:\n");
+    list_keyexchanges();
+    BIO_printf(bio_out, " Signatures:\n");
+    list_signatures();
+    BIO_printf(bio_out, " Key encapsulation:\n");
+    list_kems();
```

Comment:
Nope, these should go in `list_pkey_meth`. However, keymgmt listing should go here.

Reply:
that sounds weird - everything else is listed in the algorithm part.

Reply:
Do you really want it that way?

Reply:
In legacy, this is a bit mixed up if you look on the surface.  However, observe that the legacy bit here lists EVP_PKEY_ASN1_METHODs...  that's the bits for the key type itself, i.e. corresponds to EVP_KEYMGMT.
`list_pkey_meth()`, on the other hand, lists EVP_PKEY_METHOD, which is the collection of functions corresponding to EVP_SIGNATURE, EVP_KEYEXCH and EVP_ASYM_CIPHER combined (and EVP_KEM should be included there as well, of course).

So:

* `list_pkey()` lists key types.
* `list_pkey_meth()` lists *operation methods* done with key types.

In legacy, they look very much the same, since it's more or less a 1:1 mapping between the EVP_PKEY_ASN1_METHOD and the EVP_PKEY_METHOD, while in provider, it's a 1:*n* mapping (*n* being the amount of possible operation types for an EVP_PKEY)

Reply:
Done